### DISCUSSION TITLE
Azure Container Registry Managed Identity

### DISCUSSION THREAD

**Discussion by pietervincken** [renovatebot/renovate#24909] [Suggest an Idea]
### What would you like Renovate to be able to do?

Using a managed identity to authenticate towards the Azure Container Registry for looking up images. 

### If you have any ideas on how this should be implemented, please tell us here.

A similar approach could be used to what's [already been done for AWS.](https://github.com/joeapearson/renovate/blob/master/lib/datasource/docker/index.ts#L115) The username and password method could be preceded by a check for ACR repositories (they have a distinct domain) and could be authenticated up front through a call through the Azure SDK to detect authentication methods. 

### Is this a feature you are interested in implementing yourself?

No

**Comment by rarkins** [comment]
This needs some more research and details before it can more from requirements->ready

**Reply by bjuraga** [reply]
It's been a while. Could you please let the community know if there has been any progress on this issue?

**Comment by bjuraga** [comment]
In case anyone stumbles across this, one way to solve this without an admin user is to create a scope map on the ACR using a wildcard selector for repositories and grant the two read permissions.

Then create a token using the scope map you just created. 

Last step, create a password for that token. You get two passwords so you can rotate them and set expiration dates.

**Reply by Dutchy-** [reply]
I am in the process of figuring this out for myself, ACR scope map tokens indeed seem like the best approach without more specific renovate support.

Why do you need to create a password for that token? And how?
Renovate accepts a token for a docker registry, is it not sufficient to use the token there?

**Comment by Lingkar** [comment]
@rarkins I have done some curl experimenting and the following sequence works for me:
Does that help with understanding compatibility with the current logic?

The scope found is: repository:some/image:metadata_read

```
tenant={{TENANT_ID}}
registry={{ACR_HOST}}
client_id={{SERVICE_PRINCIPAL_ID}}
client_secret={{SERVICE_PRINCIPAL_SECRET}}
scope="https://management.azure.com//.default"
path="/v2/some/image/tags/list?n=1000"

aad_access_token=$(
    curl -X POST -H 'Content-Type: application/x-www-form-urlencoded' \
        -d "client_id=$client_id" \
        -d "grant_type=client_credentials" \
        -d "scope=$scope" \
        -d "client_secret=$client_secret" https://login.microsoftonline.com/$tenant/oauth2/v2.0/token | jq -r .access_token
)

echo "AAD access token:"
echo $aad_access_token
echo

acr_refresh_token=$(
    curl -s -X POST -H "Content-Type: application/x-www-form-urlencoded" \
        -d "grant_type=access_token&service=$registry&tenant=$tenant&access_token=$aad_access_token" \
        https://$registry/oauth2/exchange | jq -r .refresh_token
)

echo "ACR refresh token:"
echo $acr_refresh_token
echo

scope=$(curl -s https://$registry$path | jq -r '.errors[].detail[] | "\(.Type):\(.Name):\(.Action)"')

echo "Scope $scope"
echo

acr_access_token=$(
    curl -s -X POST -H "Content-Type: application/x-www-form-urlencoded" \
        -d "grant_type=refresh_token&service=$registry&scope=$scope&refresh_token=$acr_refresh_token" \
        https://$registry/oauth2/token | jq -r .access_token
)

echo "ACR access token:"
echo $acr_access_token
echo

curl -H "Authorization: Bearer $acr_access_token" https://$registry$path
```

Usefull docs:
https://github.com/Azure/acr/blob/main/docs/AAD-OAuth.md#calling-post-oauth2exchange-to-get-an-acr-refresh-token

**Reply by Lingkar** [reply]
I will give it an implementation attempt:
https://github.com/Lingkar/renovate/tree/feat/acr-service-principal-authentication

**Reply by Codex-** [reply]
Managed to encounter this a few times

@Lingkar did you get very far with this? Appreciate your efforts! 🚀

**Reply by Lingkar** [reply]
> Managed to encounter this a few times
> 
> @Lingkar did you get very far with this? Appreciate your efforts! 🚀

@Codex- The implementation took longer than expected mainly because I couldn't find a way to get the needed token from the ACR library @azure/container-registry. So instead used the @azure/identity library to get the needed token(s).
I haven't really had time to continue working on it, I believe in my fork the authentication flow is working, still had to test workload identities. 
The implementation needs an additional config variable for supplying the Azure tenant id, in my fork this is now taken from an ENV variable instead of usual Renovate config way, so that is still a TODO. And I did not yet get to the point of implementing any tests for it.
I might have some time in the upcoming weeks to try and finish it up, but not sure.
